test/extended/ha/ha.go

```diff
@@ -0,0 +1,258 @@
+	}
+	// FIXME: Controllers instance immediatelly terminates after acquiring its lease without following sleep
+	time.Sleep(1 * time.Second)
+
+	// One controllers instance needs to be running before CLI can be initialized
+	_, _err = m.StartNewInstance()
+	if _err != nil {
+		g.Fail(fmt.Sprintf("Failed to initialize controllers instance: %v\n", _err))
+	}
+	_, _, _err = m.WaitForActive(time.Second * 2)
+	if _err != nil {
+		g.Fail(_err.Error())
+	}
+
+	oc := exutil.NewCLI("controller-leases", exutil.KubeConfigPath())
+
+	// WaitForPod blocks until given pod terminates.
+	// FIXME: this is probably too complicated
+	// FIXME: consider using oc logs directly
+	WaitForPod := func(podName string, timeout time.Duration) {
+		start := time.Now()
+		cmd := exec.Command("oc", "get", "-n", oc.Namespace(), "--watch", "--no-headers", "pod", podName)
+		stdout, err := cmd.StdoutPipe()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		stderr, err := cmd.StderrPipe()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		ec := make(chan error)
+		err = cmd.Start()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		// log stderr
+		go func() {
+			io.Copy(g.GinkgoWriter, stderr)
+		}()
+
+		// log stdout and look for exited state
+		go func() {
+			bufReader := bufio.NewReader(stdout)
+			for {
+				line, err := bufReader.ReadString('\n')
+				if err != nil {
+					ec <- err
+					return
+				}
+				g.GinkgoWriter.Write([]byte(fmt.Sprintf("`oc get --watch` output: %s", line)))
+				submatch := reExitCode.FindStringSubmatch(line)
+				if len(submatch) > 1 {
+					if submatch[1] != "0" {
+						ec <- fmt.Errorf("Pod terminated with unexpected exit code (%s)", submatch[1])
+					} else {
+						ec <- nil
+					}
+					return
+				}
+			}
+		}()
+
+		select {
+		case err := <-ec:
+			if err != nil && err != io.EOF {
+				o.Expect(err).NotTo(o.HaveOccurred())
+			}
+			if err == nil {
+				cmd.Process.Kill()
+				cmd.Wait()
+			} else {
+				o.Expect(cmd.Wait()).NotTo(o.HaveOccurred())
+			}
+		case <-time.After(start.Add(timeout).Sub(time.Now())):
+			g.Fail(fmt.Sprintf("Timeout (%s) occurred while waiting for pod's exit", timeout.String()))
+		}
+	}
+
+	StartPod := func() {
+		g.By(fmt.Sprintf("\n HA TEST:  Creating a pod from %s", podConfig))
+		err := oc.Run("create").Args("-f", podConfig).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By("\n HA TEST:  Wait for pod's exit")
+		WaitForPod("helloworld-pod", time.Second*50)
+
+		g.By("\n HA TEST:  Check pod's output")
+		out, err := oc.Run("logs").Args("--interactive=false", "helloworld-pod").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(strings.Contains(strings.ToLower(out), "hello from docker"))
+	}
+
+	// EnsureOneInstanceActive releases any inactive controllers instances and
+	// makes sure that one is active.
+	EnsureOneInstanceActive := func() {
+		l := m.Len()
+		switch {
+		case l > 1:
+			g.By("\n HA TEST:  Releasing inactive controllers")
+			m.ReleaseControllers(m.GetInactive()...)
+		case l == 0:
```

Comment:
don't use `l`.  It led to me scratching my head saying, "wtf, why is he testing if one equals zero??" :)